### PR TITLE
Implement Vector{2|3|4}

### PR DESCRIPTION
**Pull request by BeLeap**
This revision includes:
- Implement Vector{2|3|4} (#15)
  - Inner product
  - Outer product
  - Scale
  - Negate
  - Squared Length
  - Get Length
  - Normalize
  - Add
  - Sub

### CODE REVIEW

File: oxide_core/test/vector/vector3.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    #[test]
+    fn normalize() {
+        let new_vector = vector3::Vector3 {
+            x: 3.0,
+            y: 4.0,
+            z: 5.0,
+        };
+        let tar_vector = vector3::Vector3 {
+            x: 0.42426407871192851464050661726291,
```
Comment (hyp3rflow) [inline oxide_core/test/vector/vector3.rs]: Due to precision limit, ```0.42426407871192851464050661726291``` evaluates to ```0.4242640787119285```.
```suggestion
            x: 0.424_264_07,
```

File: oxide_core/vector/vector2.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+#[derive(PartialEq, Debug, Serialize, Deserialize)]
+pub struct Vector2 {
+    pub x: f32,
+    pub y: f32,
+}
+
+impl vector::Rotate for Vector2 {
+    fn rotate(&self, _rotation_matrix: matrix::Matrix) -> Self {
+        Self { x: 0.0, y: 0.0 }
+    }
```
Comment (hyp3rflow) [inline oxide_core/vector/vector2.rs]:
```suggestion
    // @TODO
    fn rotate(&self, _rotation_matrix: matrix::Matrix) -> Self {
        panic!("TODO");
   	}
```

**Review by hyp3rflow — APPROVED** [review]

**Comment by hyp3rflow** [comment]
approve 잘못누름 ㅠ

**Comment by utilForever** [comment]
@hyp3rflow 그거 다시 Request changes로 바꾸면 되잖어

File: oxide_core/test/vector/vector3.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            y: 0.56568542494923801952067548968388,
```
Comment (utilForever) [inline oxide_core/test/vector/vector3.rs]: Same as above.
```suggestion
            y: 0.565_685_45,
```

File: oxide_core/test/vector/vector3.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            z: 0.70710678118654752440084436210485,
```
Comment (utilForever) [inline oxide_core/test/vector/vector3.rs]: Same as above.
```suggestion
            z: 0.707_106_77,

```

File: oxide_core/test/vector/vector4.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
+#[cfg(test)]
+mod test {
+    use crate::vector::*;
+
+    fn assert_vector4_approx_eq(a: &vector4::Vector4, b: &vector4::Vector4) {
+        assert_approx_eq!(a.x, b.x);
+        assert_approx_eq!(a.y, b.y);
+        assert_approx_eq!(a.z, b.z);
+        assert_approx_eq!(a.w, b.w);
+    }
```
Comment (utilForever) [inline oxide_core/test/vector/vector4.rs]: This code should move to `oxide_core/test/utils.rs`.

File: oxide_core/test/vector/vector4.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+    #[test]
+    fn get_length() {
+        let new_vector = vector4::Vector4 {
+            x: 3.0,
+            y: 4.0,
+            z: 5.0,
+            w: 6.0,
+        };
+        let tar: f32 = 9.2736184954957037525164160739902;
```
Comment (utilForever) [inline oxide_core/test/vector/vector4.rs]: Same as above.
```suggestion
        let tar: f32 = 9.273_619;
```

File: oxide_core/test/vector/vector4.rs [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    #[test]
+    fn normalize() {
+        let new_vector = vector4::Vector4 {
+            x: 3.0,
+            y: 4.0,
+            z: 5.0,
+            w: 6.0,
+        };
+        let tar_vector = vector4::Vector4 {
+            x: 0.32349831961031524718080521188338,
```
Comment (utilForever) [inline oxide_core/test/vector/vector4.rs]: Same as above.
```suggestion
            x: 0.323_498_3,
```

File: oxide_core/test/vector/vector4.rs [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            y: 0.43133109281375366290774028251117,
```
Comment (utilForever) [inline oxide_core/test/vector/vector4.rs]: Same as above.
```suggestion
            y: 0.431_331_1,

```

File: oxide_core/test/vector/vector4.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            z: 0.53916386601719207863467535313896,
```
Comment (utilForever) [inline oxide_core/test/vector/vector4.rs]: Same as above.
```suggestion
            z: 0.539_163_9,
```

File: oxide_core/test/vector/vector4.rs [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            w: 0.64699663922063049436161042376676,
```
Comment (utilForever) [inline oxide_core/test/vector/vector4.rs]: Same as above.
```suggestion
            w: 0.646_996_6,
```

File: oxide_core/vector/vector3.rs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+
+    fn negate(&self) -> Self {
+        self.scale(-1.0)
+    }
+}
+
+impl ops::Add<Vector3> for Vector3 {
+    type Output = Vector3;
```
Comment (utilForever) [inline oxide_core/vector/vector3.rs:65]: What is `type Output`?

File: oxide_core/vector/vector3.rs [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        Vector3 {
+            x: self.x + _rhs.x,
+            y: self.y + _rhs.y,
+            z: self.z + _rhs.z,
+        }
+    }
+}
+
+impl ops::Sub<Vector3> for Vector3 {
+    type Output = Vector3;
```
Comment (utilForever) [inline oxide_core/vector/vector3.rs:77]: Same as above.

File: oxide_core/vector/vector4.rs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        }
+    }
+
+    fn negate(&self) -> Self {
+        self.scale(-1.0)
+    }
+}
+
+impl ops::Add<Vector4> for Vector4 {
+    type Output = Vector4;
```
Comment (utilForever) [inline oxide_core/vector/vector4.rs:57]: Same as above.

File: oxide_core/vector/vector4.rs [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+            x: self.x + _rhs.x,
+            y: self.y + _rhs.y,
+            z: self.z + _rhs.z,
+            w: self.w + _rhs.w,
+        }
+    }
+}
+
+impl ops::Sub<Vector4> for Vector4 {
+    type Output = Vector4;
```
Comment (utilForever) [inline oxide_core/vector/vector4.rs:70]: Same as above.

File: oxide_core/vector/vector2.rs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    fn add(self, _rhs: Vector2) -> Vector2 {
+        Vector2 {
+            x: self.x + _rhs.x,
+            y: self.y + _rhs.y,
+        }
+    }
+}
+
+impl ops::Sub<Vector2> for Vector2 {
+    type Output = Vector2;
```
Comment (utilForever) [inline oxide_core/vector/vector2.rs:80]: Same as above.

File: oxide_core/vector/vector2.rs [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        self.get_squared_length().sqrt()
+    }
+
+    fn normalize(&self) -> Self {
+        self.scale(self.get_length().recip())
+    }
+}
+
+impl ops::Add<Vector2> for Vector2 {
+    type Output = Vector2;
```
Comment (utilForever) [inline oxide_core/vector/vector2.rs:69]: Same as above.

**Review by utilForever — CHANGES_REQUESTED** [review]
You should change some codes.
Moreover, I think that you should add unit tests for default value of vector{2|3|4}.

File: oxide_core/vector/vector3.rs (hunk 11 above)
Comment (BeLeap) [reply oxide_core/vector/vector3.rs:65]: You have to do that to do Operator Overloading in Rust.

File: oxide_core/test/vector/vector3.rs (hunk 1 above)
Comment (BeLeap) [reply oxide_core/test/vector/vector3.rs]: I know it's more than its precision, but I've been thinking about when the implementation changes in the future (for example, when the implementation changes to f64), what would be better?

File: oxide_core/vector/vector2.rs (hunk 2 above)
Comment (BeLeap) [reply oxide_core/vector/vector2.rs]: I think it's a very good idea.

File: oxide_core/test/vector/vector4.rs (hunk 5 above)
Comment (BeLeap) [reply oxide_core/test/vector/vector4.rs]: Done

File: oxide_core/test/vector/vector3.rs (hunk 1 above)
Comment (utilForever) [reply oxide_core/test/vector/vector3.rs]: If there is an occasion to change the type to "f64", I think the test code should be changed accordingly. The current type is "f32", so the test code should be tailored as well.

Comment (utilForever) [reply oxide_core/test/vector/vector3.rs]: @hyp3rflow What do you think?

Comment (hyp3rflow) [reply oxide_core/test/vector/vector3.rs]: > The current type is "f32", so the test code should be tailored as well.  

I agree with you 😄 @utilForever

Comment (LuneFleurs) [reply oxide_core/test/vector/vector3.rs]: I also agree with this change.

**Review by LuneFleurs — APPROVED** [review]

**Review by hyp3rflow — APPROVED** [review]
LGTM 🚀

**Review by utilForever — APPROVED** [review]
LGTM now. Thanks for your work! 🛩️

**Pull request merged**